scipy/stats/_distn_infrastructure.py

```diff
@@ -1252,21 +1261,25 @@ def moment(self, n, *args, **kwds):
                 mdict = {'moments': {1: 'm', 2: 'v', 3: 'vs', 4: 'vk'}[n]}
             else:
                 mdict = {}
-            mu, mu2, g1, g2 = self._stats(*args, **mdict)
-        val = _moment_from_stats(n, mu, mu2, g1, g2, self._munp, args)
+            mu, mu2, g1, g2 = self._stats(*goodargs, **mdict)
+        val = _moment_from_stats(n, mu, mu2, g1, g2, self._munp, goodargs)
 
         # Convert to transformed  X = L + S*Y
         # E[X^n] = E[(L+S*Y)^n] = L^n sum(comb(n, k)*(S/L)^k E[Y^k], k=0...n)
-        if loc == 0:
-            return scale**n * val
-        else:
-            result = 0
-            fac = float(scale) / float(loc)
-            for k in range(n):
-                valk = _moment_from_stats(k, mu, mu2, g1, g2, self._munp, args)
-                result += comb(n, k, exact=True)*(fac**k) * valk
-            result += fac**n * val
-            return result * loc**n
+        fac = _lazywhere(cond1_, (goodloc, goodscale),
+                         lambda loc, scale: scale / loc, self.badvalue)
```

Comment:
Since we're still adding `nan`s here where `loc==0`, we still do operations on those elements below.  Ideally, we wouldn't be touching the elements where `loc == 0` until we get to  `res = goodscale**n * val`. And ideally, that statement shouldn't touch the elements where `loc != 0`. I think the solution would be to do another `argsreduce` after you get `val` for all the `goodargs`.

I'll submit a PR to your branch so you can see what I mean.

Reply:
Thanks! I merged it from my fork.